File: app/src/main/java/org/ohosdev/hapviewerandroid/adapter/InfoAdapter.java
package org.ohosdev.hapviewerandroid.adapter;

import android.annotation.SuppressLint;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.snackbar.Snackbar;

import org.ohosdev.hapviewerandroid.R;
import org.ohosdev.hapviewerandroid.model.HapInfo;

public class InfoAdapter extends RecyclerView.Adapter<InfoAdapter.ViewHolder> {

    private static final int MAX_ITEMS = 6;
    private final LayoutInflater layoutInflater;
    private final String unknownString;
    @NonNull
    private HapInfo info = new HapInfo();

    public InfoAdapter(Context context) {
        super();
        layoutInflater = LayoutInflater.from(context);
        unknownString = context.getString(android.R.string.unknownName);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(layoutInflater.inflate(R.layout.info_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {

        switch (position) {
            case 0:
                holder.setName(R.string.info_appName);
                holder.setContent(info.init ? unknownString : info.appName);
                break;
            case 1:
                holder.setName(R.string.info_appPackageName);
                holder.setContent(info.init ? unknownString : info.packageName);
                break;
            case 2:
                holder.setName(R.string.info_versionName);
                holder.setContent(info.init ? unknownString : info.versionName);
                break;
            case 3:
                holder.setName(R.string.info_versionCode);
                holder.setContent(info.init ? unknownString : info.versionCode);
                break;
            case 4:
                holder.setName(R.string.info_targetAPI);
                holder.setContent(info.init ? unknownString : String.format("API %s (%s)", info.targetAPIVersion, info.apiReleaseType));
                break;
            case 5:
                holder.setName(R.string.info_tech);
                holder.setContent(info.init ? unknownString : info.getTechDesc(holder.context));
                break;
        }
    }

    @Override
    public int getItemCount() {
        return MAX_ITEMS;
    }

    @SuppressLint("NotifyDataSetChanged")
    public void setInfo(@NonNull HapInfo info) {
        this.info = info;
        notifyDataSetChanged();
    }


    static class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private final TextView textView;
        private final Context context;
        @NonNull
        private String name = "";
        @NonNull
        private String content = "";

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            textView = itemView.findViewById(R.id.textView);
            itemView.setOnClickListener(this);
            context = itemView.getContext();
        }

        private void refresh() {
            textView.setText(String.format("%s: %s", name, content));
        }

        public void setName(@Nullable String name) {
            if (name == null || name.equals(this.name))
                return;
            this.name = name;
            refresh();
        }

        public void setName(int resId) {
            setName(context.getString(resId));
        }

        public void setContent(@Nullable String content) {
            if (content == null || content.equals(this.content))
                return;
            this.content = content;
            refresh();
        }

        public void copyText() {
            if (content.isEmpty())
                return;

            ClipboardManager cm = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
            // 设置内容到剪切板
            cm.setPrimaryClip(ClipData.newPlainText(null, this.content));
            // Toast.makeText(itemView.getContext(), "已复制 " + name, Toast.LENGTH_SHORT).show();
            String toastText = String.format(context.getString(R.string.copied_withName), name);
            Snackbar.make(itemView, toastText, Snackbar.LENGTH_SHORT).show();
        }

        @Override
        public void onClick(View v) {
            copyText();
        }
    }
}

File: app/src/main/java/org/ohosdev/hapviewerandroid/adapter/InfoAdapter.kt
package org.ohosdev.hapviewerandroid.adapter

import android.annotation.SuppressLint
import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import org.ohosdev.hapviewerandroid.R
import org.ohosdev.hapviewerandroid.app.BaseActivity
import org.ohosdev.hapviewerandroid.databinding.ItemInfoBinding
import org.ohosdev.hapviewerandroid.extensions.getTechDesc
import org.ohosdev.hapviewerandroid.model.HapInfo


class InfoAdapter(
    val context: BaseActivity,
    private val onItemClick: (holder: InfoAdapter.ViewHolder) -> Unit
) : RecyclerView.Adapter<InfoAdapter.ViewHolder>() {
    private val layoutInflater: LayoutInflater = context.layoutInflater
    private val unknownString: String = context.getString(android.R.string.unknownName)
    private var info = HapInfo()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(ItemInfoBinding.inflate(layoutInflater, parent, false), onItemClick)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        when (position) {
            0 -> holder.bind(R.string.info_appName, info.appName)
            1 -> holder.bind(R.string.info_appPackageName, info.packageName)
            2 -> holder.bind(R.string.info_versionName, info.versionName)
            3 -> holder.bind(R.string.info_versionCode, info.versionCode)
            4 -> holder.bind(
                R.string.info_compileTarget, "API ${info.targetAPIVersion} (${info.apiReleaseType})"
            )

            5 -> holder.bind(
                R.string.info_tech,
                info.getTechDesc(context) ?: context.getString(R.string.info_tech_unknown)
            )
        }
    }

    override fun getItemCount(): Int {
        return MAX_ITEMS
    }

    @SuppressLint("NotifyDataSetChanged")
    fun setInfo(info: HapInfo) {
        this.info = info
        notifyDataSetChanged()
    }

    inner class ViewHolder(
        private val binding: ItemInfoBinding,
        onItemClick: (holder: InfoAdapter.ViewHolder) -> Unit
    ) : RecyclerView.ViewHolder(binding.root) {

        var name = ""
        var content = ""

        init {
            binding.root.setOnClickListener { onItemClick(this) }
        }

        @SuppressLint("SetTextI18n")
        private fun refresh() {
            binding.textView.text = "${name}: $content"
        }

        private fun setName(resId: Int) {
            name = context.getString(resId)
        }

        fun bind(nameId: Int, content: String?) {
            setName(nameId)
            this.content = if (info.init || content == null) unknownString else content
            refresh()
        }
    }

    companion object {
        private const val MAX_ITEMS = 6
    }
}